quickfix.c: cannot specify a count with :chistory, vim-patch:8.1.{1241,1261}, 8.2.{0095,0113}

erw7 opened the pull request:
Fixes #13076.

#### vim-patch:8.1.1241: Ex command info contains confusing information

Problem:    Ex command info contains confusing information.
Solution:   When using the NOTADR flag use ADDR_OTHER for the address type. Cleanup code using NOTADR.  Check for errors in create_cmdidxs.vim.  Adjust Makefile to see the errors.
vim/vim@b731689

#### vim-patch:8.1.1261: no error for quickfix commands with negative range

Problem:    No error for quickfix commands with negative range.
Solution:   Add ADDR_UNSIGNED and use it for quickfix commands.  Make assert_fails() show the command if the error doesn't match.
vim/vim@25190db

#### vim-patch:8.2.0095: cannot specify exit code for :cquit

Problem:    Cannot specify exit code for :cquit.
Solution:   Add optional argument. (Thinca, Yegappan Lakshmanan, closes vim/vim#5442)
vim/vim@1860bde

#### vim-patch:8.2.0113: "make cmdidxs" fails

Problem:    "make cmdidxs" fails.
Solution:   Allow address for ":cquit".  Add --not-a-term to avoid a delay.
vim/vim@9b24dfc

erw7 committed:
quickfix.c: cannot specify a count with :chistory

janlazo commented on src/nvim/ex_cmds.lua:
```diff
@@ -483,8 +483,8 @@ return {
-    flags=bit.bor(TRLBAR),
-    addr_type=ADDR_LINES,
+    flags=bit.bor(RANGE, COUNT, TRLBAR),
+    addr_type=ADDR_OTHER,
```
This should be `ADDR_UNSIGNED` which comes from patch v8.1.1261. Is this PR a short term fix for #13076 ?

erw7 committed:
vim-patch:8.1.1261: no error for quickfix commands with negative range

Problem:    No error for quickfix commands with negative range.
Solution:   Add ADDR_UNSIGNED and use it for quickfix commands.  Make
            assert_fails() show the command if the error doesn't match.
https://github.com/vim/vim/commit/25190db225d63e185e77e043e694ef455b3cf304

erw7 committed:
vim-patch:8.1.1241: Ex command info contains confusing information

Problem:    Ex command info contains confusing information.
Solution:   When using the NOTADR flag use ADDR_OTHER for the address type.
            Cleanup code using NOTADR.  Check for errors in
            create_cmdidxs.vim.  Adjust Makefile to see the errors.
https://github.com/vim/vim/commit/b731689e85b4153af7edc8f0a6b9f99d36d8b011

janlazo commented:
Patch 8.1.1210, 8.1.1241 should go first before patch 8.1.1261. It's fine if you can make patch 8.1.1261 pass without them.

erw7 committed:
vim-patch:8.2.0095: cannot specify exit code for :cquit

Problem:    Cannot specify exit code for :cquit.
Solution:   Add optional argument. (Thinca, Yegappan Lakshmanan, closes vim/vim#5442)
https://github.com/vim/vim/commit/1860bde9d31bbb0ba857f6284f6332a7134030dd

erw7 committed:
vim-patch:8.2.0113: "make cmdidxs" fails

Problem:    "make cmdidxs" fails.
Solution:   Allow address for ":cquit".  Add --not-a-term to avoid a delay.
https://github.com/vim/vim/commit/9b24dfcb9f676e7f7a09a9062f0d05b2104a87eb

erw7 committed:
test, legacy/assert_spec.lua: fix few tests fail

Adjust failing tests to current behavior due to changes in assert_fails
behavior by fbd6a86.

erw7 committed:
test, provider/define_spec.lua: fix few tests fail

fbd6a86 changes the behavior of :command-range and no longer allows
arbitrary numbers. Adjust tests to reflect this behavior.

janlazo commented:
What's the status of this PR? Any blockers?

erw7 commented:
> What's the status of this PR? Any blockers?

I don't merge changes into the master that have not been approved by anyone, except minor ones. I'm waiting for someone to review it. 

There doesn't seem to be any rules for this project, so maybe it's okay to merge them if they pass the test.

janlazo commented:
Ok, I'll review.

janlazo commented on src/nvim/generators/gen_ex_cmds.lua:
```diff
@@ -62,9 +66,24 @@ for _, cmd in ipairs(defs) do
+    .cmd_addr_type = %s
   },
-]], enumname, cmd.command, cmd.func, cmd.flags, cmd.addr_type))
+  ]], enumname, cmd.command, cmd.func, cmd.flags, cmd.addr_type))
```
Extra whitespace?

janlazo commented on src/nvim/ex_cmds.lua:
```diff
@@ -15,3288 +15,3277 @@ local COUNT      =    0x400
+local CMDWIN     =  0x80000
+local MODIFY     = 0x100000
+local EXFLAGS    = 0x200000
+local RESTRICT   = 0x400000
```
https://github.com/neovim/neovim/pull/11996 removed restricted mode so `RESTRICT` is now dead code. Maybe comment it out and update affected ex commands.

janlazo commented on src/nvim/ex_cmds_defs.h:
```diff
@@ -51,32 +51,36 @@
+    ADDR_QUICKFIX_VALID,  // quickfix list valid entry number
+    ADDR_QUICKFIX,        // quickfix list entry number
+    ADDR_UNSIGNED,        // positive count or zero, defaults to 1
+    ADDR_OTHER,           // something else
```
Patch 8.1.1261

`ADDR_OTHER,		 // something else, use line number for '$',`

janlazo commented:
```
 /*
  * Find an Ex command by its name, either built-in or user.
  * Start of the name can be found at eap->cmd.
- * Returns pointer to char after the command name.
+ * Sets eap->cmdidx and returns a pointer to char after the command name.
  * "full" is set to TRUE if the whole command name matched.
  * Returns NULL for an ambiguous user command.
  */
     static char_u *
 find_command(exarg_T *eap, int *full UNUSED)
```

janlazo commented:
WIP on reviewing ex_cmds.lua because of the `addr_type` changes.

janlazo commented:
I ported 8.2.0095. vim/vim#5442 mentions that it is based on Neovim implementation so porting it should not require changes in the C,lua code.
